File: apps/api-rs/src/conversation_title.rs
use std::sync::Arc;
use std::time::Duration;

use tracing::warn;
use uuid::Uuid;

use crate::agent::model::{AgentMessage, AgentModelRequest};
use crate::agent::AgentModel;
use crate::models::{MessageRole, MessageStatus};
use crate::repositories::ConversationRepository;

const MAX_TITLE_CHARS: usize = 10;
const UPDATE_INTERVAL: usize = 4;
const FIRST_RECURRING_UPDATE: usize = 3;
const RECENT_TURNS: usize = 3;
const TITLE_TIMEOUT: Duration = Duration::from_secs(12);

const TITLE_SYSTEM_PROMPT: &str = r#"你是专业的会话标题生成器。请为对话生成简洁的中文标题。

规则：
1. 标题不超过 10 个字
2. 提炼核心主题或用户意图
3. 使用名词短语
4. 不使用标点、引号或书名号
5. 只输出标题，不要解释"#;

pub fn spawn_title_update(
    repository: Arc<dyn ConversationRepository>,
    model: Arc<dyn AgentModel>,
    tenant_id: Uuid,
    user_id: Uuid,
    conversation_id: Uuid,
) -> tokio::task::JoinHandle<Option<String>> {
    tokio::spawn(async move {
        match generate_and_update_title(&repository, model, tenant_id, user_id, conversation_id)
            .await
        {
            Ok(title) => title,
            Err(error) => {
                warn!(
                    %conversation_id,
                    error = %error,
                    "asynchronous conversation title generation failed"
                );
                None
            }
        }
    })
}

async fn generate_and_update_title(
    repository: &Arc<dyn ConversationRepository>,
    model: Arc<dyn AgentModel>,
    tenant_id: Uuid,
    user_id: Uuid,
    conversation_id: Uuid,
) -> anyhow::Result<Option<String>> {
    let session = match repository.get_session(tenant_id, conversation_id).await? {
        Some(session) if session.user_id == user_id => session,
        _ => return Ok(None),
    };
    let messages = repository.get_messages(tenant_id, conversation_id).await?;
    let user_message_count = messages
        .iter()
        .filter(|message| message.role == MessageRole::User)
        .count();
    if !should_generate_title(user_message_count) {
        return Ok(None);
    }
    if user_message_count == 1 && session.title.trim() != "新会话" {
        return Ok(None);
    }

    let prompt = if user_message_count == 1 {
        let first_message = messages
            .iter()
            .find(|message| message.role == MessageRole::User)
            .map(|message| truncate_chars(&message.content, 500))
            .unwrap_or_default();
        format!(
            "请为以下用户消息生成一个 10 字以内的中文标题：\n\n用户消息：\n{first_message}\n\n请直接输出标题："
        )
    } else {
        format!(
            "请根据以下对话生成一个 10 字以内的中文标题：\n\n{}\n\n请直接输出标题：",
            recent_conversation(&messages)
        )
    };

    let request = AgentModelRequest {
        messages: vec![
            AgentMessage::system(TITLE_SYSTEM_PROMPT),
            AgentMessage::user(prompt),
        ],
        tools: Vec::new(),
        temperature: 0.2,
        max_tokens: 32,
    };
    let response = tokio::time::timeout(TITLE_TIMEOUT, model.complete(request)).await??;
    let Some(title) = response.content.as_deref().and_then(normalize_title) else {
        return Ok(None);
    };
    let updated = repository
        .update_session_title(tenant_id, user_id, conversation_id, &title, false)
        .await?;
    Ok(updated.then_some(title))
}

fn should_generate_title(user_message_count: usize) -> bool {
    user_message_count == 1
        || (user_message_count >= FIRST_RECURRING_UPDATE
            && (user_message_count - FIRST_RECURRING_UPDATE) % UPDATE_INTERVAL == 0)
}

fn recent_conversation(messages: &[crate::models::message::ConversationMessage]) -> String {
    let mut recent = Vec::new();
    let mut user_count = 0usize;
    for message in messages.iter().rev().filter(|message| {
        message.role == MessageRole::User
            || (message.role == MessageRole::Assistant
                && message.status == MessageStatus::Completed)
    }) {
        recent.push(message);
        if message.role == MessageRole::User {
            user_count += 1;
            if user_count == RECENT_TURNS {
                break;
            }
        }
    }
    recent.reverse();

    recent
        .into_iter()
        .map(|message| {
            let role = if message.role == MessageRole::User {
                "用户"
            } else {
                "助手"
            };
            format!("{role}：{}", truncate_chars(&message.content, 200))
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn normalize_title(raw: &str) -> Option<String> {
    let first_line = raw.lines().next()?.trim();
    let clean = first_line
        .trim_matches(|character| {
            matches!(
                character,
                '"' | '\'' | '“' | '”' | '‘' | '’' | '《' | '》' | '。' | '！' | '？'
            )
        })
        .trim();
    if clean.is_empty() {
        return None;
    }
    Some(truncate_chars(clean, MAX_TITLE_CHARS))
}

fn truncate_chars(value: &str, limit: usize) -> String {
    value.chars().take(limit).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn moss_title_update_cadence_is_preserved() {
        let due = (0..12)
            .filter(|count| should_generate_title(*count))
            .collect::<Vec<_>>();
        assert_eq!(due, vec![1, 3, 7, 11]);
    }

    #[test]
    fn title_is_single_line_unquoted_and_ten_chars() {
        assert_eq!(
            normalize_title("“企业知识库检索性能优化”\n解释").as_deref(),
            Some("企业知识库检索性能优")
        );
    }
}

File: apps/api-bun/src/conversation_title.ts
// 移植自 apps/api-rs/src/conversation_title.rs
import { completePiText, type PiModelSettings } from './agent/pi/model.ts';
import type { ConversationRepository } from './repositories/types.ts';
import type { ConversationMessage } from './models/message.ts';

const MAX_TITLE_CHARS = 10;
const UPDATE_INTERVAL = 4;
const FIRST_RECURRING_UPDATE = 3;
const RECENT_TURNS = 3;
const TITLE_TIMEOUT_MS = 12_000;

const TITLE_SYSTEM_PROMPT = `你是专业的会话标题生成器。请为对话生成简洁的中文标题。

规则：
1. 标题不超过 10 个字
2. 提炼核心主题或用户意图
3. 使用名词短语
4. 不使用标点、引号或书名号
5. 只输出标题，不要解释`;

export function spawnTitleUpdate(
  repository: ConversationRepository,
  settings: PiModelSettings,
  tenantId: string,
  userId: string,
  conversationId: string,
): Promise<string | null> {
  return generateAndUpdateTitle(repository, settings, tenantId, userId, conversationId)
    .catch((error: unknown) => {
      console.warn(`[documind][title] conversation ${conversationId} title generation failed: ${(error as Error).message}`);
      return null;
    });
}

async function generateAndUpdateTitle(
  repository: ConversationRepository,
  settings: PiModelSettings,
  tenantId: string,
  userId: string,
  conversationId: string,
): Promise<string | null> {
  const session = await repository.getSession(tenantId, conversationId);
  if (!session || session.user_id !== userId) return null;
  const messages = await repository.getMessages(tenantId, conversationId);
  const userMessageCount = messages.filter((message) => message.role === 'user').length;
  if (!shouldGenerateTitle(userMessageCount)) return null;
  if (userMessageCount === 1 && session.title.trim() !== '新会话') return null;

  const prompt = userMessageCount === 1
    ? (() => {
        const firstMessage = messages
          .find((message) => message.role === 'user');
        const content = firstMessage ? truncateChars(firstMessage.content, 500) : '';
        return `请为以下用户消息生成一个 10 字以内的中文标题：\n\n用户消息：\n${content}\n\n请直接输出标题：`;
      })()
    : `请根据以下对话生成一个 10 字以内的中文标题：\n\n${recentConversation(messages)}\n\n请直接输出标题：`;

  const fallback = normalizeTitle(
    messages.find((message) => message.role === 'user')?.content ?? '',
  ) ?? '新会话';
  let timer: NodeJS.Timeout | undefined;
  let title = fallback;
  try {
    const response = await Promise.race([
      completePiText(
        { ...settings, temperature: 0.2, maxTokens: 32, thinkingEnabled: false },
        TITLE_SYSTEM_PROMPT,
        prompt,
      ),
      new Promise<never>((_, reject) => {
        timer = setTimeout(
          () => reject(new Error('title generation timed out')), TITLE_TIMEOUT_MS);
      }),
    ]);
    title = normalizeTitle(response) ?? fallback;
  } catch (error) {
    console.warn(
      `[documind][title] conversation ${conversationId} model generation failed, using fallback: ${(error as Error).message}`,
    );
  } finally {
    clearTimeout(timer);
  }
  const updated = await repository.updateSessionTitle(
    tenantId, userId, conversationId, title, false);
  return updated ? title : null;
}

export function shouldGenerateTitle(userMessageCount: number): boolean {
  return userMessageCount === 1
    || (userMessageCount >= FIRST_RECURRING_UPDATE
      && (userMessageCount - FIRST_RECURRING_UPDATE) % UPDATE_INTERVAL === 0);
}

function recentConversation(messages: ConversationMessage[]): string {
  const recent: ConversationMessage[] = [];
  let userCount = 0;
  const candidates = messages.filter((message) =>
    message.role === 'user'
    || (message.role === 'assistant' && message.status === 'completed'));
  for (const message of candidates.reverse()) {
    recent.push(message);
    if (message.role === 'user') {
      userCount += 1;
      if (userCount === RECENT_TURNS) break;
    }
  }
  recent.reverse();
  return recent
    .map((message) => {
      const role = message.role === 'user' ? '用户' : '助手';
      return `${role}：${truncateChars(message.content, 200)}`;
    })
    .join('\n');
}

export function normalizeTitle(raw: string): string | null {
  const firstLine = raw.split(/\r?\n/)[0];
  if (firstLine === undefined) return null;
  const clean = firstLine.trim()
    .replace(/^["'“”‘’《》。！？、：:；;]+|["'“”‘’《》。！？、：:；;]+$/gu, '')
    .replace(/^(?:标题|会话标题)\s*[:：]\s*/u, '')
    .replace(/^["'“”‘’《》。！？、：:；;]+/u, '')
    .replace(/^(?:请帮我|帮我|请问|请|麻烦)\s*/u, '')
    .replace(/^(?:总结|分析|统计|查询|列出|说明|介绍|查看|对比)\s*/u, '')
    .replace(/["'“”‘’《》。！？、：:；;]+$/u, '')
    .replace(/\s+/gu, ' ')
    .trim();
  if (clean.length === 0) return null;
  return truncateChars(clean, MAX_TITLE_CHARS);
}

function truncateChars(value: string, limit: number): string {
  return [...value].slice(0, limit).join('');
}
